feat(agw): Added code to support dedicated bearer establishment for roaming UE

rsarwad opened the pull request:
feat(agw): Added code to support dedicated bearer establishment for roaming UE

## Summary
feat(agw): Added code to support dedicated bearer establishment for roaming UE

## Test Plan
Executed s1ap sanity test suite and tested dedicated bearer creation for roaming UE in Tera VM setup

rsarwad committed:
Added code to support dedicated bearer for roaming UEs

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

uri200 committed:
[feg] s8_proxy split createBearerRequest into request and response

Signed-off-by: Oriol Batalla <obatalla@fb.com>
(cherry picked from commit 0e6cd65f1f49b3222e36c154c802bc8a5197941b)

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Added dedicated bearer support for roaming UEs

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Provided fixes for the issues found during testing of dedicated bearer establishment

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Corrected number of packet filters

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Added debug logs

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Cleaned up the code

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

ulaskozat commented on lte/gateway/c/core/oai/tasks/sgw/pgw_handlers.c:
```diff
@@ -628,39 +620,38 @@ static int create_temporary_dedicated_bearer_context(
   // Create temporary spgw bearer context entry
   pgw_ni_cbr_proc_t* pgw_ni_cbr_proc =
-      pgw_get_procedure_create_bearer(spgw_ctxt_p);
+      pgw_get_procedure_create_bearer(sgw_ctxt_p);
```
although nit really part of this PR, could you take care of github action comment for line 630 and replace `strcpy` with `snprintf`

ulaskozat commented on lte/gateway/c/core/oai/tasks/sgw/sgw_handlers.c:
```diff
@@ -1919,29 +1915,31 @@ bool is_enb_ip_address_same(const fteid_t* fte_p, ip_address_t* ip_p) {
-              policy_rule_name,
-              sgw_eps_bearer_entry_p->sgw_eps_bearer_entry->policy_rule_name);
+          if (module == LOG_SPGW_APP) {
+            strcpy(
```
legit github action comment about strcpy.

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

pshelar committed:
fix(mme): handling of dedicated bearer for inbound roaming.

This PR fixes typo in dedicated bearer related event in GTP APP
and adds unit test to validate flow generation.

Signed-off-by: Pravin B Shelar <pbshelar@fb.com>

rsarwad committed:
Modified code to update flow precedence while adding OVS rules for S8 tunnel belonging to dedicated bearer

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

uri200 commented on lte/gateway/c/core/oai/lib/s8_proxy/s8_client_api.cpp:
```diff
@@ -529,3 +544,62 @@ void send_s8_delete_session_request(
+    magma::feg::CreateBearerResponsePgw* proto_cb_rsp, teid_t pgw_s8_teid,
+    uint32_t sequence_number, char* pgw_cp_address, Imsi_t imsi) {
+  OAILOG_FUNC_IN(LOG_SGW_S8);
+  proto_cb_rsp->Clear();
```
just curious, why do we need to clear the content of proto_cb_rsp here?

rsarwad replied:
It's just to clear the contents before filling the proto structure

uri200 replied:
but what is the reason `proto_cb_rsp` may have any content ? isn't it passed as brand new object?

or is this function just setting default values (like ints to 0)

rsarwad replied:
Ya, the purpose of clear is to set back to empty state. Since it's new object we can remove it.

uri200 commented on lte/gateway/c/core/oai/tasks/grpc_service/S8ServiceImpl.h:
```diff
@@ -0,0 +1,46 @@
+      Void* response) override;
+};
+
+}  // namespace magma
```
once we figure out the s6a_task new grpc service, we may need to refactor this to use the same idea, instead of breaking the calls into two steps (request and answer)

rsarwad replied:
Created an issue, https://github.com/magma/magma/issues/9157 for tracking

uri200 commented on lte/gateway/c/core/oai/tasks/mme_app/mme_app_bearer.c:
```diff
@@ -271,12 +278,21 @@ status_code_e send_pcrf_bearer_actv_rsp(
+  if (pdn_context->route_s11_messages_to_s8_task) {
+    OAILOG_INFO_UE(
+        LOG_MME_APP, ue_context_p->emm_context._imsi64,
+        "Sending create_dedicated_bearer_rsp to SGW_s8 task for EBI %u s1u "
```
is `SGW_s8` the standard name we use to identify that task in other log files?

uri200 replied:
Should we add the IMSI besides the EBI on the log so we can identify what IMSI is being created. Or is the IMSI printed somewhere else during this process? or is `MME_UE_S1AP_ID_FMT` the IMSI itself ?

rsarwad replied:
Imsi is logged as part of OAILOG_INFO_UE.

rsarwad replied:
> is `SGW_s8` the standard name we use to identify that task in other log files?

Create bearer response is sent from MME; So included LOG_MME_APP is included.
If any handling is performed in context of SGW_S8 task, while logging LOG_SGW_S8 is used.

uri200 commented on lte/gateway/c/core/oai/test/openflow/test_gtp_app.cpp:
```diff
@@ -1014,6 +1014,150 @@ TEST_F(GTPApplicationTest, TestDeleteTunnelS8) {
+
+  controller->dispatch_event(del_tunnel);
+}
+
```
👍

rsarwad committed:
Incorporated review comments

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Resolved merge conflicts

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

ulaskozat commented on lte/gateway/c/core/oai/tasks/grpc_service/S8ServiceImpl.cpp:
```diff
@@ -0,0 +1,106 @@
+    s8_create_bearer_request_t* itti_msg) {
+  auto ip                  = request->pgwaddrs();
+  itti_msg->pgw_cp_address = (char*) calloc(1, (ip.size() + 1));
+  strcpy(itti_msg->pgw_cp_address, ip.c_str());
```
legit github action comment

ulaskozat commented on lte/gateway/configs/service_registry.yml:
```diff
@@ -110,3 +110,6 @@ services:
     port: 50084
+  s8_service:
+    ip_address: 127.0.0.1
+    port: 50073
```
why do you use the same port as SPGW service?

rsarwad replied:
As part of grpc initialization, grpc server is created with specific ip and port number. And there are multiple services like s1ap, sgs, ha etc are getting registred to this server. Similarly I have registered s8 service to existing server with same ip and port number.
Here is the link for initializing grpc server.
https://github.com/magma/magma/blob/master/lte/gateway/c/core/oai/tasks/grpc_service/grpc_service.cpp#L62

ulaskozat replied:
ah indeed, you are reusing the grpc server task. then can you please place this together with 50073 services?

ulaskozat commented:
Folks, there is absolutely no unit test coverage for any of this. What is the plan?

rsarwad committed:
Incorporated review comments

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Incorporated review comments

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

pshelar committed:
fix(mme): pgw api for dedicated bearer

This patch removes dependency of flow-dl for
dedicated bearer of inbound roaming UEs. This also
simplifies the inbound roaming flows by matching on
PGW TEIDs

Signed-off-by: Pravin B Shelar <pbshelar@fb.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
For failed create bearer response, getting bearer context for failed bearer and fetching sequence number and pgw_cp_ip

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Fixed clang errors

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Fixed the compilation

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Fixed issues found when UE rejects dedicatd bearer creation

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Fixed compilation errors occurring on OAI build

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Added unit test cases for dedicated bearer creation for roaming UEs

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Fixed clang errors

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Added code to set port to ovs switch

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
iving absolute path for ovs-vsctl commands

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Added code remove initialization open flow controller

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Addressed self review comments

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

ulaskozat reviewed (approved):
LGTM. Please also check for any memory leaks reported on syslog by running the teravm test and then restarting the services.

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad commented:
> LGTM. Please also check for any memory leaks reported on syslog by running the teravm test and then restarting the services.

Hi Ulas, 
There are other memory leaks observed during configuration, I think we can take as separate activity

rsarwad committed:
Fixed memory leak

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>

rsarwad committed:
Merge remote-tracking branch 'upstream/master' into rsarwad_dedicated_bearer_roam

Signed-off-by: Rashmi <rashmi.sarwad@radisys.com>
